### PR TITLE
Proposed Opentracing TCK Core Implementation

### PR DESCRIPTION
**Pull request by kgibm**
This PR proposes the core of the opentracing TCK implementation for discussion. Highlights:

- Assumes that the application server under test provides a [MockTracer](https://github.com/opentracing/opentracing-java/tree/master/opentracing-mock) to the test application. Based on the opentracing spec, this is `@Inject`ed into the `Tracer` in the JAXRS `org.eclipse.microprofile.opentracing.tck.application.TestWebServices` class. In addition to a "Hello World" JAXRS endpoint that will be tested, this class has a method named `getTracer` which will be called by the Arquillian client to return the information stored in the `MockTracer` so that the client can assert the expected opentracing spans. In addition, this class has a method named `clearTracer` which can be called in between tests to clear the `MockTracer`.
- The method `TestWebServices.getTracer` uses reflection to access the information of the `MockTracer`. The reason this is done instead of bringing in the `io.opentracing:opentracing-mock` dependency directly and casting the `Tracer` to `MockTracer` is that we don't want to assume that the application's classloader has visibility into the classloader that created the `Tracer` (if there is no such visibility, a `ClassCastException` will occur).
- The method `TestWebServices.getTracer` converts the data in the `MockTracer` into a `TestTracer` and `TestSpan`s which abstracts out the logical information in the `MockTracer` (this is also needed because `MockTracer` is not a POJO so its data is not transparently serialized by `Jackson`).
- The class `org.eclipse.microprofile.opentracing.tck.OpentracingClientTests` is the `Arquillian` test class. In the `createDeployment` method, we generate the WAR that gets installed into the target application server.
- There is a single test, `simpleTest` which has the `@RunAsClient` annotation meaning it will run in the `Arquillian` test harness and act as the JAXRS client. In `executeHelloWorld`, we get get the remote `TestTracer` and call its `spanTree` method which returns a `TestSpanTree`. This converts all of the spans into a tree data structure which allows for easier comparison (to assert equality) and also overrides `toString` to print out a JSON-like representation of the tree of spans (if the equality assertion fails, it's nice to compare the trees to see what's different).
- Finally we build a simulated `TestSpanTree` of the spans that we expect the response to give, and we assert equality between the two. The `TestSpanTree` and its tree of `TestSpan`s override `equals` to recursively perform the comparison based on the number of spans at each level and the spans' operation name and span kind attributes.

Signed-off-by: Kevin Grigorenko <kevin.grigorenko@us.ibm.com>

### CODE REVIEW

File: tck/running_the_tck.asciidoc [hunk 1]
```diff
@@ -22,6 +22,8 @@ The TCK uses `testng`
 
 == Dependencies
 
+The application server must make a MockTracer (https://github.com/opentracing/opentracing-java/tree/master/opentracing-mock) available to the application.
+
 To enable the tests in your project you need to add the following dependency to your build:
 
 [source, xml]
```
Comment (kwsutter) [inline tck/running_the_tck.asciidoc]: Do you really want this type of detail in the spec document?  Looks like something that could easily get out of sync with the actual pom.xml files...

File: tck/src/main/java/org/eclipse/microprofile/opentracing/tck/OpentracingClientTests.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2017 Contributors to the Eclipse Foundation
```
Comment (kwsutter) [inline tck/src/main/java/org/eclipse/microprofile/opentracing/tck/OpentracingClientTests.java]: Maybe you did this already, but I haven't seen an update to the NOTICE file to indicate the additional contributions by KevinG...  Any others?

File: tck/pom.xml [hunk 3]
```diff
@@ -45,58 +45,74 @@
...
-		    <groupId>org.jboss.arquillian.junit</groupId>
-		    <artifactId>arquillian-junit-container</artifactId>
-		</dependency>
-		<dependency>
-		    <groupId>junit</groupId>
-		    <artifactId>junit</artifactId>
-		    <version>4.12</version>
-		    <scope>test</scope>
-		</dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (kgibm) [reply tck/pom.xml]: Based on the spreadsheet at that link, I changed `javax.ws.rs-api` and `jackson-jaxrs-json-provider` to already approved CQs, and the ones remaining are `io.opentracing:opentracing-api:0.30.0`, `org.jboss.shrinkwrap.resolver:shrinkwrap-resolver-impl-maven:2.2.4`, `org.jboss.resteasy:resteasy-client:3.1.4.Final`, and `org.jboss.resteasy:resteasy-jackson-provider:3.1.4.Final`. I tried version `3.0.19.Final` of RESTEasy which has a CQ but it threw an errors. The client only needs to be a simple JAXRS POJO client, so we can also look at already approved alternate JAXRS clients.

File: tck/running_the_tck.asciidoc (hunk 1 above)
Comment (kgibm) [reply tck/running_the_tck.asciidoc]: Agreed, I'll do more research and potentially ask the mailing list, although for now it should be okay as it matches mpConfig.

File: tck/src/main/java/org/eclipse/microprofile/opentracing/tck/OpentracingClientTests.java (hunk 2 above)
Comment (kgibm) [reply tck/src/main/java/org/eclipse/microprofile/opentracing/tck/OpentracingClientTests.java]: Updated

**Comment by kgibm** [comment]
Ready to be merged.

**Pull request merged**